BHoM type vs Revit family name mapping implemented, parameter mapping on filtering enabled + BuildersWork.Opening convert added

pawelbaran opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

### NOTE: Depends on 
<!-- Link to any additional PRs in other repos required for this PR to function -->
<!-- Delete if not required -->
https://github.com/BHoM/BHoM_Datasets/pull/94

   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1054
Half of #1012

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
Minimal, problem-specific test is located [on SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRevit%5FToolkit%2F%231060%2DNEWTypeAndParameterMapping&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).
Standard tests are available [here](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F03%5FAlpha%2FBHoM%2FRevit%5FToolkit&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4) (mainly _Pull_ and _Requests_ should be interesting), with the test procedure explained [here](https://burohappold.sharepoint.com/:w:/r/sites/BHoM/_layouts/15/Doc.aspx?sourcedoc=%7BC1165A69-0022-420D-BD0E-CE46F3C50A79%7D&file=0008_Revit_Toolkit.docx&action=default&mobileredirect=true).


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->
- mechanism of choosing the applicable convert method on Pull refined and linked with the mechanism of filtering by type
- BHoM type vs Revit family name mapping on pull implemented
- parameter mapping unlocked on filtering by parameter
- `FromRevit` convert added for `Architecture.BuildersWork.Opening`

### Additional comments
<!-- As required -->
The code of this PR is ready for review, regardless of the failing checks. Since the PR is touching the core aspects of the adapter, extensive tests will be required - I have just started testing, would be great if the reviewers could try to stress and break it too.

Key information on the refactoring that happened (see TL;DR section if you do not care about the details 😃):
- Choice of the convert method to call does not happen through dynamic casting anymore (`FromRevit(element as dynamic)` is gone). Instead, following procedure occurs, on very high level:
    1. Find the target BHoMType for given Revit element, `Discipline` and adapter settings using `BHoMType` method - this method took over all switch/cases of disciplines that previously sat in `FromRevit`
    2. Find relevant Revit => BHoM convert method for the target BHoM type using `ConvertMethod` query method
    3. Call this method if found, if anything goes wrong then convert to `ModelInstance`
    4. The convert is finalised with `Postprocess` method that currently contains only transformations, but open to further extension, e.g. removal of calls to parameter/identifier copying in each individual convert method and moving them there
- `BHoMType` can be called from anywhere, therefore can be used to refine filtering by type, incl. taking family mapping into account.
- Being able to determine the output type before the actual convert (using `BHoMType` method) unlocked the ability to apply parameter mapping on filtering by parameter.
- Family name mapping is kept separate from parameter mapping because the former applies only to chosen types on Pull, while the latter is a general concept used across the adapter. Mixing the two could cause lots of confusion plus it would be harder to implement.

A flowchart showing the sequence of calls together with some comments:

![mapping](https://user-images.githubusercontent.com/26874773/128010460-f1257ad1-9c21-40c2-9deb-432e1ce667e0.png)

### TL;DR:
This PR changes the inner mechanics of the adapter and enables family name mapping on pull as well as parameter mapping on filtering by parameter. Family mapping only applies to the types that make use of it inside `BHoMType` method, currently the only one is `BH.oM.Architecture.BuildersWork.Opening`, also included in this PR.

I know this is a monster of PR, but I think it introduces really valuable changes and puts a lot of order into how the adapter works. I will continue refining the descriptions, null handling etc., so please do not merge. `ParameterSettings` to be renamed to `MappingSettings` in a separate PR.

FraserGreenroyd commented:
@BHoMBot check core
@BHoMBot check compliance

pawelbaran commented:
@BHoMBot check null-handling

pawelbaran commented:
@BHoMBot check versioning

pawelbaran commented:
@BHoMBot check serialisation
@BHoMBot check installer

pawelbaran commented:
Looks like only the descriptions are missing, so happy to get the code reviewed too, if anyone fancies reading through over 1000 lines 😉

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check documentation-compliance

pawelbaran commented:
@BHoMBot check required

pawelbaran commented:
/azp run Revit_Toolkit.CheckCore

pawelbaran commented:
@enarhi @tiagogrossi I believe the PR is ready for a full review - I did finalise the descriptions, ran the tests, bugfixed and benchmarked:
- the new convert method selection mechanism seems to work fine for all test cases
- filtering by type is slower than on `master`, but this is the price for robustness and clarity that enabled mapping of family names against BHoM types

The changes are quite massive, but it would be really hard to split them - I hope you will not find them impossible to review 😉

FraserGreenroyd commented:
> I've reviewed the functionality so far, it's looking good. I think we discussed the workflow enough and I'm happy with what you came up with. One small request I have is to enable `RecordError` somewhere when the family name is not found. I made a typo when I tested and all I got was this clueless msg.
> ![image](https://user-images.githubusercontent.com/20777038/127981832-e2761193-dd06-4c0c-bea8-606d12fe0b3d.png)

What does the note contain that's attached to the component? I would say an error would be misleading - it might be deliberate that there are no family names in the model for what you've searched for. An error suggests something went wrong rather than a valid return. A warning might be appropriate but I'd rather it be a note because effectively it is an FYI rather than a HELP response.

pawelbaran commented:
> > I've reviewed the functionality so far, it's looking good. I think we discussed the workflow enough and I'm happy with what you came up with. One small request I have is to enable `RecordError` somewhere when the family name is not found. I made a typo when I tested and all I got was this clueless msg.
> > ![image](https://user-images.githubusercontent.com/20777038/127981832-e2761193-dd06-4c0c-bea8-606d12fe0b3d.png)
> 
> What does the note contain that's attached to the component? I would say an error would be misleading - it might be deliberate that there are no family names in the model for what you've searched for. An error suggests something went wrong rather than a valid return. A warning might be appropriate but I'd rather it be a note because effectively it is an FYI rather than a HELP response.

I am afraid this would be tricky to implement @tiagogrossi: if we are meant to raise a warning every time there is no valid family map for a Revit element, you would end up with it being raised for the elements you do not want to pull, i.e. imagine you have 2 families of Generic Components in your model, you map one of them into openings, but the other will keep complaining that there is no valid map for it, which you know and intend. Or do you see a smarter approach to it?

tiagogrossi reviewed (commented):
Did a first run in the code and have some silly questions. As mentioned offline it would be greatly beneficial to have a visual organogram, starting from the very start (from read?) till the converted obj.

FraserGreenroyd commented on Revit_Core_Engine/Convert/Physical/FromRevit/Construction.cs:
```diff
@@ -35,6 +36,24 @@ public static partial class Convert
 
+        [Description("Converts a Revit HostObjAttributes to an Physical.Constructions.Construction.")]
+        [Input("hostObjAttributes", "Revit HostObjAttributes to be converted.")]
+        [Input("settings", "Revit adapter settings to be used while performing the convert.")]
```
That would be different to the convention done by the rest of BHoM, so for consistency I would not agree.

pawelbaran replied:
I have never thought of that tbh. But agreed with @FraserGreenroyd in this case, I think the general (unwritten?) convention in BHoM is not to mention that a parameter is optional.

However, maybe it would make sense to refine the `Description` query to check for the default value and mention it if true?

tiagogrossi replied:
OKe, good to know!

tiagogrossi commented:
> 
> 
> > > I've reviewed the functionality so far, it's looking good. I think we discussed the workflow enough and I'm happy with what you came up with. One small request I have is to enable `RecordError` somewhere when the family name is not found. I made a typo when I tested and all I got was this clueless msg.
> > > ![image](https://user-images.githubusercontent.com/20777038/127981832-e2761193-dd06-4c0c-bea8-606d12fe0b3d.png)
> > 
> > 
> > What does the note contain that's attached to the component? I would say an error would be misleading - it might be deliberate that there are no family names in the model for what you've searched for. An error suggests something went wrong rather than a valid return. A warning might be appropriate but I'd rather it be a note because effectively it is an FYI rather than a HELP response.
> 
> I am afraid this would be tricky to implement @tiagogrossi: if we are meant to raise a warning every time there is no valid family map for a Revit element, you would end up with it being raised for the elements you do not want to pull, i.e. imagine you have 2 families of Generic Components in your model, you map one of them into openings, but the other will keep complaining that there is no valid map for it, which you know and intend. Or do you see a smarter approach to it?

Well the user must be informed somewhere. Would perhaps `BH.Engine.Adapters.Revit.Query.MappedFamilyNames` be the right place?

pawelbaran commented on Revit_Core_Engine/Convert/Architecture/FromRevit/Opening.cs:
```diff
@@ -0,0 +1,112 @@
+            if (profile == null)
+                BH.Engine.Reflection.Compute.RecordWarning($"The profile of the opening could not be extracted from the Revit element because the parameters correspondent to their dimensions could not be found. Revit ElementId: {instance.Id}"
+                    + "\nTo link the specific parameter values with opening height/width/diameter, add relevant ParameterMap to RevitSettings.ParameterSettings.");
+
```
I always found it hard to embed Wiki links in the messages because one can't click on them. On the other hand, the link can be copied or simply browsed, so I a agree, makes sense. Issue raised: https://github.com/BHoM/Revit_Toolkit/issues/1061

pawelbaran commented on Revit_Core_Engine/Convert/Structure/FromRevit/MaterialFragment.cs:
```diff
@@ -37,6 +39,24 @@ public static partial class Convert
+        [Output("material", "Structure.MaterialFragments.IMaterialFragment resulting from converting the input Revit Material.")]
+        public static IMaterialFragment MaterialFragmentFromRevit(this Material material, RevitSettings settings = null, Dictionary<string, List<IBHoMObject>> refObjects = null)
+        {
+            return material.MaterialFragmentFromRevit(null, settings, refObjects);
```
Yeah agreed this looks a bit artificial. The reason why I added the new overloads without `grade` parameter is because the `ConvertMethod` query searches only through methods that have the following:
- return type of `IBHoMObject` or a collection of `IBHoMObjects`
- 1st input parameter of type `Element`
- 2nd input parameter of type `RevitSettings`
- 3rd input parameter of type `Dictionary<string, List<IBHoMObject>>`

This is needed to make sure that on invocation of the method the number of passed parameters is equal to the number of inputs. I will draw up a flowchart very soon to visualise the whole process.

pawelbaran commented on Revit_Core_Engine/Query/CoordinateSystem.cs:
```diff
@@ -0,0 +1,61 @@
+        [Description("Finds the coordinate system of a given Revit FamilyInstance.")]
+        [Input("familyInstance", "Revit FamilyInstance to be queried for its coordinate system.")]
+        [Output("coordinateSystem", "Coordinate system of the input Revit FamilyInstance.")]
+        public static Cartesian CoordinateSystem(this FamilyInstance familyInstance)
```
I think this is slightly different. The query you mention extracts the orientation in the BHoM convention for 1-dimensional objects, while this is a dumb query extracting the coordinate system of 0-dimensional ones. But the question is very valid, it sounds like there could be some convergence at least - to be investigated.

pawelbaran commented:
> Well the user must be informed somewhere. Would perhaps `BH.Engine.Adapters.Revit.Query.MappedFamilyNames` be the right place?

What would you inform the user about @tiagogrossi? Let's imagine you have 3 `FamilyInstances` of category X in your Revit model, only two of which are meant to be converted to BHoM based on the family mapping:
- element 1 that you want to pull/convert
- element 2 that you want to pull/convert
- element 3 that you want to rule out

You create maps for elements 1 and 2, where you make a typo for element 2. Now let's consider the two approaches: without warning (current) and with a warning when mapping is wrong (proposed):
a) no warning:
    - element 1 gets pulled and converted as expected
    - element 2 fails silently
    - element 3 keeps silent and gets skipped as expected
b) with warning:
    - element 1 gets pulled and converted as expected
    - element 2 fails and a warning is raised
    - element 3 get a warning raised even though it is meant to be simply ruled out

Neither of the above is ideal I would say, and I can't think of a way of raising the warning/error only in case the mapping is intended but failed - how would one distinguish between the intended and unintended lacks of mapping?

pawelbaran commented:
@tiagogrossi I have just added a high level flowchart with extra comments to the PR description.

tiagogrossi commented:
> 
> 
> @tiagogrossi I have just added a high level flowchart with extra comments to the PR description.

Thanks for the flowchart, it helped clarifying things. I have no further comments on the functionality and on the code. Will wait approve to see if you can work out the warning we spoke of. Curious the see @enarhi review :)

pawelbaran commented:
@tiagogrossi I have just added the family map usage check as discussed + updated the test file. Feel free to have a look! 👍

pawelbaran commented:
@BHoMBot check compliance

pawelbaran commented:
@BHoMBot check required

enarhi reviewed (approved):
LGTM. Not much more to add, just happy to have this in place, improvements from both the functionality and code side.

pawelbaran commented:
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

pawelbaran commented:
@BHoMBot check ready-to-merge

pawelbaran commented:
@BHoMBot check ready-to-merge
